### PR TITLE
BUG: login: show error message from api

### PR DESCRIPTION
**Pull request by ghost**
the API can return a more detailed error message, so show that one.

fixes certtools/intelmq-api#35

### CODE REVIEW

**Comment by monoidic** [comment]
I wonder if it'd be better to have the API return something like
```json
{"error": "Invalid username and/or password"}
```
instead of just the string? Checking for errors would just mean checking for the existence of the `error` key and the API response  would consistently be an object rather than "sometimes an object with the correct results, sometimes a string with an error message".

File: intelmq_manager/static/js/static.js [hunk 1]
```diff
@@ -526,8 +526,8 @@ $(document).ready(function() {
                 $('#modalLoginForm').modal('hide');
                 updateLoginStatus();
             } else {
-                // If authentication failed, this error message is displayed.
-                $('#loginErrorField').text('Invalid username and/or password');
+                // If authentication failed, the returned error message is displayed.
+                $('#loginErrorField').text(data);
```
Comment (monoidic) [inline intelmq_manager/static/js/static.js]:
```suggestion
                $('#loginErrorField').text(data.error);
```
This, plus changing the check above determining the request was successful from whether the `login_token` and `username` attributes are present to checking for the `error` attribute.

Comment (ghost) [reply intelmq_manager/static/js/static.js]: Hm, but the absence of the error attribute does not imply the existence of the login_token and username attributes, as something else could have gone wrong too. (or the version of the API is <= 3.0.0)

**Comment by ghost** [comment]
I've also updated certtools/intelmq-api#36 to return a dict with the error attribute

**Comment by ghost** [comment]
I just noticed that if the password is not given, the server response is
```
{"errors":{"password":"Required parameter 'password' not supplied"}}
```

~~This comes from the flask session storage module.~~ It comes from hug

That format can't be handled by the frontend at all

**Comment by monoidic** [comment]
Just adding an object to a string like that gets you a nice and meaningful `[object Object]` :P


> I just noticed that if the password is not given, the server response is
> 
> ```
> {"errors":{"password":"Required parameter 'password' not supplied"}}
> ```
> 
> ~This comes from the flask session storage module.~ It comes from hug
> 
> That format can't be handled by the frontend at all

So, we'd end up with hug errors and custom errors. Maybe throwing an exception in `login` with the error string would consistently produce hug errors?

Handling that format should be far from impossible.

**Comment by ghost** [comment]
> So, we'd end up with hug errors and custom errors. Maybe throwing an exception in `login` with the error string would consistently produce hug errors?

The inconsistency is also the return code. Our own checks have 200, hug sends 400. I'll try to get something working

File: intelmq_manager/static/js/static.js (hunk 1 above)
Comment (monoidic) [reply intelmq_manager/static/js/static.js]: I'd leave it up to the API to return the correct response when it claims it does, and treat it as a bug in intelmq-api when it doesn't uphold that claim in any way.
> (or the version of the API is <= 3.0.0)

The components should all be kept in sync, should they not? I do not think there is currently any expectation of the API being compatible with different versions of clients, forwards or backwards.

**Comment by ghost** [comment]
I don't have the capabilities to get this consistent on a short term. On the API side we would need to create custom responses with JSON-encoded response and a status code 400. On the manager side, we need to handle various types of response types.
I have some initial success, but that's not good JavaScript style. The current ideas exceed my initial intention by far and thus I fear this has to wait.

**Comment by ghost** [comment]
What I now pushed fixes the bug and even shows other errors, that's what I have as minimal solution without any refactoring of the API's inconsitent behaviour.

File: intelmq_manager/static/js/static.js [hunk 2]
```diff
@@ -525,9 +525,24 @@ $(document).ready(function() {
...
-                $('#loginErrorField').text('Invalid username and/or password');
+                // Other error, display the response for easier debugging.
+                $('#loginErrorField').text("Login failed, server response was " + data);
+            }
+        })
+        .fail(function(jqXHR, textStatus) {
+            if (typeof jqXHR.responseJSON !== 'undefined' && typeof jqXHR.responseJSON.errors !== 'undefined') {
+                let concatenated = "";
+                for (let key in jqXHR.responseJSON.errors) {
+                    console.log(jqXHR.responseJSON.errors)
```
Comment (monoidic) [inline intelmq_manager/static/js/static.js]: Maybe only print the errors once if there are multiple?
```suggestion
                console.log(jqXHR.responseJSON.errors);
                for (let key in jqXHR.responseJSON.errors) {
```

**Comment by monoidic** [comment]
I'll try to look at making the API error responses more consistent at some point to simplify this. Though it should work for now.

File: intelmq_manager/static/js/static.js (hunk 2 above)
Comment (ghost) [reply intelmq_manager/static/js/static.js]: oops, that's a debugging leftover. but using it for the unhandled error response might be a good idea

**Comment by ghost** [comment]
> I'll try to look at making the API error responses more consistent at some point to simplify this. Though it should work for now.

Thanks for your commitment :)

**Pull request merged**